tests/test-sdi-progress-window.c

```diff
@@ -0,0 +1,318 @@
+}
+
+static void describe_test(TestData *test) {
+  g_assert_nonnull(test);
+  g_assert_cmpint(test->test_number, !=, -1);
+  g_autofree gchar *text = g_strdup_printf(
+      "Test %d; Title: %s\n\nDescription: %s\n\nActions: %s", test->test_number,
+      test->title, test->description, test->actions);
+  gtk_label_set_label(window_text, text);
+}
+
+static gboolean set_progress_bar(gchar *snap_name) {
+  static guint done_tasks = 0;
+  guint total_tasks = 10;
+
+  g_autofree gchar *description =
+      g_strdup_printf("Description for task %d", done_tasks);
+  sdi_progress_window_update_progress(progress_window, snap_name, NULL,
+                                      description, done_tasks, total_tasks,
+                                      FALSE, NULL);
+  done_tasks++;
+  if (done_tasks > total_tasks)
+    done_tasks = 0;
+  return G_SOURCE_CONTINUE;
+}
+
+static void wait_for_click() {
+  loop = g_main_loop_new(NULL, FALSE);
+  g_main_loop_run(loop);
+  g_main_loop_unref(loop);
+}
+
+static void enable_buttons(gpointer data) {
+  gtk_widget_set_sensitive(GTK_WIDGET(yes_button), TRUE);
+  gtk_widget_set_sensitive(GTK_WIDGET(no_button), TRUE);
+}
+
+static void disable_buttons_for_seconds(gint seconds) {
+  gtk_widget_set_sensitive(GTK_WIDGET(yes_button), FALSE);
+  gtk_widget_set_sensitive(GTK_WIDGET(no_button), FALSE);
+  g_timeout_add_once(seconds * 1000, enable_buttons, NULL);
+}
+
+static void show_progress_window(gchar *snap_name, gchar *desktop_file) {
+  g_autoptr(GDesktopAppInfo) app_info = g_desktop_app_info_new(desktop_file);
+  g_assert_nonnull(app_info);
+  g_autofree gchar *icon = g_desktop_app_info_get_string(app_info, "Icon");
+  sdi_progress_window_begin_refresh(
+      progress_window, snap_name,
+      (gchar *)g_app_info_get_display_name(G_APP_INFO(app_info)), icon, NULL);
+}
+
+static int count_progress_childs() {
+  GtkWindow *window =
+      GTK_WINDOW(sdi_progress_window_get_window(progress_window));
+  // to differentiate between window existing and zero elements, and no window
+  if (window == NULL)
+    return -1;
+  GtkWidget *container = gtk_window_get_child(window);
+  g_assert_true(GTK_IS_BOX(container));
+  GtkWidget *child = gtk_widget_get_first_child(container);
+  int counter = 0;
+  while (child != NULL) {
+    counter++;
+    child = gtk_widget_get_next_sibling(child);
+  }
+  return counter;
+}
+
+static int count_hash_childs() {
+  GHashTable *table = sdi_progress_window_get_dialogs(progress_window);
+  g_assert_nonnull(table);
+  return g_hash_table_size(table);
+}
+
+// these are the actual tests
+
+static void test_progress_bar(TestData *test) {
+  describe_test(test);
+  show_progress_window("A-SNAP", *app_list);
+
+  g_assert_cmpint(count_hash_childs(), ==, 1);
+  g_assert_cmpint(count_progress_childs(), ==, 1);
+  gint timeout_id = g_timeout_add(500, (GSourceFunc)set_progress_bar, "A-SNAP");
+  wait_for_click();
+  g_source_remove(timeout_id);
```

Comment:
These are good cleanups, but in this case it's not needed to remove the source, I'd just use `g_timeout_add_once` and live with it.

Since you don't care about tracking what's happening once the callback is done.

Reply:
The point is that the `set_progress_bar()`  callback is called several times. I can't use `g_timeout_add_once` because, in that case, the progress bar in the window would update only once, which is not what it should do.